### PR TITLE
Llama

### LINKED ISSUE #20: Implement Llama 3 model variants
Official implementation of Llama models.

### PR DESCRIPTION
**Pull request by laggui**
Bringing the first official Llama implementation to Burn! With pre-trained weights in mpk format (hosted on HF hub).

- [x] Llama decoder-only transformer architecture
- [x] Llama 3
  - [x] Tiktoken tokenizer
  - [x] Pre-trained weights
- [x] TinyLlama
  - [x] Sentencepiece tokenizer
  - [x] Pre-trained weights

Currently the top-p sampling is done on CPU before decoding since Burn is missing categorical distribution sampling. We could improve that once everything else is done.

Closes #20

### CODE REVIEW

**Comment by laggui** [comment]
Tested with wgpu and tch (gpu). I think this is ready for review!

TinyLlama results on my dev machine:

**`Wgpu`**
```
Loading record...
Loaded in 20s
Processing prompt: How many helicopters can a human eat in one sitting?
> It's impossible to know for certain how many helicopters a human can eat in one sitting. However, it's generally accepted that humans have a limited appetite and can only eat a small amount of food at a time.

50 tokens generated (3.5432 tokens/s)

Generation completed in 0m14s
```

**`LibTorch<f16>`**
```
Loading record...
Loaded in 18s
Processing prompt: How many helicopters can a human eat in one sitting?
> It's impossible to know for certain how many helicopters a human can eat in one sitting. However, it's generally accepted that humans have a limited appetite and can only eat a small amount of food at a time.

50 tokens generated (21.6305 tokens/s)

Generation completed in 0m2s
```

Pretty big difference 😅

File: llama-burn/src/sampling.rs [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        });
+        // .scan(0.0, |acc, (idx, x)| {
+        //     *acc = *acc + x;
+        //     // Clamp smaller probabilities to zero.
+        //     if *acc >= self.p {
+        //         probs_sort[idx] = 0.0.elem();
+        //     }
+        //     Some(*acc)
+        // })
+        // .collect::<Vec<_>>();
```
Comment (nathanielsimard) [inline llama-burn/src/sampling.rs:85]: Do we have those operations?

**Review by nathanielsimard — APPROVED** [review]
LGTM, I have only one small comment, but otherwise very good job! 👏

File: llama-burn/src/sampling.rs (hunk 1 above)
Comment (laggui) [reply llama-burn/src/sampling.rs:85]: Not yet. There's a draft PR for `cumsum` I believe but we also need to sample from a multinomial which we don't have.

That's why the current implementation only sorts the tensor and performs the rest on CPU. Would have to time it to see what the impact is.

**Comment by laggui** [comment]
Weights have been updated to use the named mpk format (much faster now that data is treated as bytes with serde). In follow-up PRs we will add quantization and support for Llama 3.1.

**Review by nathanielsimard — APPROVED** [review]

**Pull request merged**